### PR TITLE
Respect the passed dtype with `self.log`

### LINKED ISSUE #8930: Strange visualization of the log during the test
## 🐛 Bug

Hello, 

in module 'test_epoch_end' I calculate the final metric

```py
score = self.metric(valid_preds, valid_labels)

print(score)

self.log("Final_metric", score)

logging_info('Test end Ok!')
```

but after getting the output, I found that the accuracy of the data written to the log differs from the original value

## Result

console output:
```
0.8403041825095058
Test end Ok!

DATALOADER:0 TEST RESULTS
{'Final_metric': 0.8403041958808899}
```

score 0.8403041825095058 != 0.8403041958808899

I think I can draw the wrong conclusion from such an error in the results?

pytorch-lightning==1.4.1
torch==1.9.0

### LINKED ISSUE #9713: logging precision issue
### To Reproduce

https://github.com/PyTorchLightning/pytorch-lightning/issues/8887
https://github.com/PyTorchLightning/pytorch-lightning/issues/8930

This problem is logging precision. 
If I use float32, logging process change my logging
(ex: .
validation_epoch_end(self, metrics)
 'val_clk_sum': tensor(3211., device='cuda:0'),
-->
trainer.logged_metrics
 'val_clk_sum': tensor(3210.9998, device='cuda:0'),
)

I know that if I use float 64 this problem doesn't happen.
But I don't want to use the trainer default of 64 because of logging.

### Expected behavior

Above mentioned

### Environment

- PyTorch Lightning Version (e.g., 1.3.0):
- PyTorch Version (e.g., 1.8)
- Python version:
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by qqueing**
## What does this PR do?

Author: @carmocca

Fixes #9713
Fixes #8930

### Does your PR introduce any breaking changes? If yes, please list them.

Batch size must be fixed point. This would only be breaking if somebody was importing and using `ResultMetric`s directly (please don't)

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Comment by tchaton** [comment]
@carmocca Any updates on this PR ?

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 1]
```diff
@@ -185,13 +185,13 @@ def __init__(self, metadata: _Metadata, is_tensor: bool) -> None:
...
-            self.add_state("value", torch.tensor(0, dtype=torch.float), dist_reduce_fx=torch.sum)
+            self.add_state("value", torch.tensor(0, dtype=torch.float64), dist_reduce_fx=torch.sum)
             if self.meta.is_mean_reduction:
-                self.add_state("cumulated_batch_size", torch.tensor(0, dtype=torch.float), dist_reduce_fx=torch.sum)
+                self.add_state("cumulated_batch_size", torch.tensor(0, dtype=torch.float64), dist_reduce_fx=torch.sum)
 
     def update(self, value: _IN_METRIC, batch_size: torch.Tensor) -> None:
         if self.is_tensor:
-            value = value.float()
+            value = value.type(torch.float64)
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:194]: I don't want to force everybody to use full precision, I think my proposal is good enough, where you can set the precision of your tensors before logging them.

**Review by carmocca — APPROVED** [review]
@tchaton I finished it.

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 2]
```diff
@@ -207,13 +207,22 @@ def __init__(self, metadata: _Metadata, is_tensor: bool) -> None:
...
+                self.add_state("cumulated_batch_size", torch.tensor(0), dist_reduce_fx=torch.sum)
 
     def update(self, value: _IN_METRIC, batch_size: torch.Tensor) -> None:
         if self.is_tensor:
-            value = value.float()
+            if not torch.is_floating_point(value):
+                dtype = torch.get_default_dtype()
+                warning_cache.warn(
+                    # do not include the value to avoid cache misses
+                    f"You called `self.log({self.meta.name}, ...)` in your `{self.meta.fx}` but the value needs to"
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:221]:
```suggestion
                    f"You called `self.log({self.meta.name!r}, ...)` in your `{self.meta.fx}` but the value needs to"
```

File: tests/core/test_metric_result_integration.py [hunk 3]
```diff
@@ -549,12 +549,40 @@ def on_train_epoch_end(self) -> None:
...
+    metadata.sync = _Sync()
+    rm = ResultMetric(metadata, is_tensor=True)
+
+    assert rm.value.dtype == floating_dtype
+    assert rm.cumulated_batch_size.dtype == fixed_dtype
+
+    # two fixed point numbers - should be converted
+    value, batch_size = torch.tensor(2), torch.tensor(3)
+    assert value.dtype == fixed_dtype
+    with pytest.warns(UserWarning, match=rf"`self.log\(bar, ...\)` in your `foo` .* Converting it to {floating_dtype}"):
```
Comment (rohitgr7) [inline tests/core/test_metric_result_integration.py:577]:
```suggestion
    with pytest.warns(UserWarning, match=rf"`self.log\('bar', ...\)` in your `foo` .* Converting it to {floating_dtype}"):
```

**Review by rohitgr7 — APPROVED** [review]
nit

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:221]: nit !

**Review by tchaton — APPROVED** [review]
LGTM ! Great work !
